rust/lance-encoding/src/encodings/physical/basic.rs

```diff
@@ -167,51 +169,68 @@ struct BasicPageDecoder {
-        // The validity buffer is only required if we have some nulls
-        buffers[0].1 = matches!(self.mode, DataNullStatus::Some(_));
-        if let Some(values) = self.mode.values_decoder() {
-            values.update_capacity(rows_to_skip, num_rows, &mut buffers[1..], all_null);
-        } else {
-            *all_null = true;
-        }
-    }
+    ) -> Result<Vec<BytesMut>> {
+        if let Some(_values) = self.mode.values_decoder() {
+            match &self.mode {
+                DataNullStatus::Some(decoders) => {
+                    let mut dest_buffers =
+                        decoders
+                            .validity
+                            .decode_into(rows_to_skip, num_rows, all_null)?; // buffer 0
+                    let mut values_bytesmut =
+                        decoders
+                            .values
+                            .decode_into(rows_to_skip, num_rows, all_null)?; // buffer 1 onwards
 
-    fn decode_into(
-        &self,
-        rows_to_skip: u32,
-        num_rows: u32,
-        dest_buffers: &mut [bytes::BytesMut],
-    ) -> Result<()> {
-        match &self.mode {
-            DataNullStatus::Some(decoders) => {
-                decoders
-                    .validity
-                    .decode_into(rows_to_skip, num_rows, &mut dest_buffers[..1])?;
-                decoders
-                    .values
-                    .decode_into(rows_to_skip, num_rows, &mut dest_buffers[1..])?;
-            }
-            // Either dest_buffers[0] is empty, in which case these are no-ops, or one of the
-            // other pages needed the buffer, in which case we need to fill our section
-            DataNullStatus::All => {
-                dest_buffers[0].fill(0);
-            }
-            DataNullStatus::None(values) => {
-                dest_buffers[0].fill(1);
-                values.decode_into(rows_to_skip, num_rows, &mut dest_buffers[1..])?;
+                    dest_buffers.append(&mut values_bytesmut);
+                    Ok(dest_buffers)
+                }
+                // Either dest_buffers[0] is empty, in which case these are no-ops, or one of the
+                // other pages needed the buffer, in which case we need to fill our section
+                DataNullStatus::All => {
+                    let mut capacities = vec![(0, false); self.num_buffers() as usize];
+
+                    // No need to look at the validity decoder to know the dest buffer size since it is boolean
+                    capacities[0].0 = arrow_buffer::bit_util::ceil(num_rows as usize, 8) as u64;
+                    // The validity buffer is only required if we have some nulls
+                    capacities[0].1 = matches!(self.mode, DataNullStatus::Some(_));
+
+                    let mut dest_buffers = create_buffers_from_capacities(capacities);
+
+                    dest_buffers[0].fill(0);
+
+                    Ok(dest_buffers)
+                }
```

Comment:
This is unreachable.  If we are in this mode then `self.mode.values_decoder` is `None` in which case we will not be in this if branch.  Is this code important?  If so, it probably belongs in the `else` branch (although the whole top-level if/else branch should just go away per previous comment)